simpsom/network.py

```diff
@@ -455,28 +457,37 @@ def train(self, train_algo: str = "batch", epochs: int = -1,
             # and parallelization at the cost of memory.
             # The object-oriented structure is kept to simplify code reading.
 
-            all_weights = self.xp.array([n.weights for n in self.nodes_list])
-            all_weights = all_weights.reshape(
-                self.width, self.height, self.data.shape[1])
+            all_weights = self.xp.array([n.weights for n in self.nodes_list], dtype=self.xp.float32)
+            all_weights = all_weights.reshape(self.width, self.height, self.data.shape[1])
 
             numerator = self.xp.zeros(all_weights.shape, dtype=self.xp.float32)
             denominator = self.xp.zeros(
                 (all_weights.shape[0], all_weights.shape[1], 1), dtype=self.xp.float32)
 
-            unravel_precomputed = self.xp.unravel_index(self.xp.arange(self.width * self.height),
-                                                        (self.width, self.height))
+            unravel_precomputed = self.xp.unravel_index(self.xp.arange(self.width * self.height, dtype=self.xp.int64), (self.width, self.height))
 
-            _xx, _yy = self.xp.meshgrid(self.xp.arange(
-                self.width), self.xp.arange(self.height))
+            _xx, _yy = self.xp.meshgrid(self.xp.arange(self.width), self.xp.arange(self.height))
 
-            neighborhood_caller = partial(
-                self.neighborhoods.neighborhood_caller, xx=_xx, yy=_yy,
-                neigh_func=self.neighborhood_fun,
-                pbc_func=self.polygons.neighborhood_pbc if self.PBC
-                else None)
+            # neighborhood_caller = partial(
+            #     self.neighborhoods.neighborhood_caller, xx=_xx, yy=_yy,
+            #     neigh_func=self.neighborhood_fun,
+            #     pbc_func=self.polygons.neighborhood_pbc if self.PBC
+            #     else None)
+
+            if self.PBC:
+                pbc_func_params = self.polygons.neighborhood_pbc
+            else:
+                pbc_func_params = None
+
+            self.neighborhoods = Neighborhoods(self.xp, _xx, _yy, pbc_func_params)
```

Comment:
If the idea is to have Neighborhoods instantiated as a part of each training then it shouldn't be set as an attribute to the network class. It can be a local variable.